File: src/BotPhysics.ts
import BotPos from "./models/BotPos";
import EnginePower from "./models/EnginePower";

export default interface BotPhysics {
    start(onUpdate: PositionCallback, initialPosition: readonly BotPos[]): void;
    setPower(power: EnginePower[]): void;
}

export interface PositionCallback{
    (this: void, positions: BotPos[]): void
}

File: putlebot-server/src/bots/bot_physics.rs
use crate::bots::engine_power::EnginePower;
use futures::Sink;
use crate::bots::bot_pos::BotPos;

pub trait BotPhysics {
    fn start(on_update: dyn Sink<Vec<BotPos>, Error = ()>,  initial_position: &[BotPos]);
    fn set_power(power: &[EnginePower]);
}